### ISSUE TITLE
[Accessibility] There should be a way to get out of the CodeMirror section with the keyboard.

### ISSUE THREAD

**Issue by yzen** [devtools-html/debugger.html#4076]
Right now if keyboard focus lands inside the code mirror section, there is no way to get out of it which in turn creates a keyboard trap. 

Perhaps a good approach is to be able to press ESC that would shift focus on the code mirror container so the consequent tabs/shift+tabs will result in navigation to the next/previous sections outside of the code mirror section.

Also it would be nice to not automatically step into code when navigating with the keyboard and instead focus on the container. To step into code, the user could press Enter/Space to activate. 

The mouse behaviour should remain the same.

**Comment by jasonLaster** [comment]
this is an interesting problem. The pattern with devtools is that escape should open the console. Infact, i think most people would consider it a bug if we don't do that. Unfortunately, that takes moves focus pretty far out of the current context.

**Comment by Garbee** [comment]
Also, escape is problematic if say a tooltip is open. As escape should close the tooltips in the editor without removing focus. Escaping within CM is not a clear-cut thing as to what expected results should be.

<kbd><kbd>Shift</kbd>+</kbd>Tab</kbd></kbd> currently has no context in CodeMirror. So we could hijack this as our escape hatch. But, I'd much prefer we try to collaborate with this with the Chrome DevTools and Edge teams as well. This affects all of us from what I can tell and we should all find a consistent way for everyone to implement so this is stable across toolsets. If we do one key combo and another toolset decides `esc` is in fact the proper solution, then we're making a space complex for users that really need these that shouldn't be.

**Comment by yzen** [comment]
Some thoughts:
The problem with code mirror is twofold, imo. 
Firstly, for efficiency, tabbing to the CodeMirror area should not move the focus straight into the the editable space. The focus should remain on the container and editing state should be triggered by user action (e.g. pressing Enter/Space). If user activates code mirror, focus moves inside as it currently does. If the user carries on with tabbing/shift-tabbing, the focus moves forward/back without stepping inside.
Secondly, the keyboard trap itself. Originally I proposed Escape because this is what we settled on when working on keyboard accessibility for our DevTools Inspector. If the state of the inspector is in edit mode, escape reverts it to read only mode and only the second escape triggers the console. So the precedent is already there, for several releases, without much negative feedback, afaik.
Alternatively, a shortcut that could potentially work could be Ctrl[Cmd]+Tab for tabbing forward away from the editable area and Shift+Ctrl[Cmd]+Tab for tabbing back. This however must be discoverable in some sort of way since it's not something the user can intuitively figure out.

**Comment by WaltersJesse** [comment]
/unclaim
My team members and I have not solved the bug entirely, but we do have some thoughts and possibly a place to start looking.

To attempt to fix the bug, we changed code in two files: 
In src/utils/editor/editor/index.js we made the following changes:
we specified a keybind (Ctrl+Shift+Tab) to enter and exit the CodeMirror when navigating with the keyboard.

In /src/components/Editor/index.js we made the following changes: 
we modified the onKeyDown function at line 283 with a conditional that will handle our keybind (Ctrl + Shift + Tab for the sake of simplicity). The new conditional looks like: if (key ==="Ctrl + Shift + Tab" && target == textArea) and inside we called the stopPropogation function 

we modified the componentDidMount function in this file and added a line: shortcuts.on("Ctrl + Shift+Tab", this.ourFunction);

We also had changes that we did not get a chance to implement:
In src/utils/editor/editor/index.js:
We will need to modify the extraKeys dictionary that is initialized in the createEditor function in index.js and add the keybind.

In /src/components/Editor/index.js: 
We planned to call a function to shift focus to the source list pane. This function would shift focus back into the source list pane when the keybind is activated. The function would be located in this file and be similar to onEscape, but for the keybind. 

Here is a git diff showing our changes.

![capture](https://user-images.githubusercontent.com/32164453/33810318-4b107424-ddd1-11e7-9bf1-418ed3c5f671.PNG)

**Comment by jasonLaster** [comment]
I'm moving this to the icebox, we'll re-evaluate in the next release process. 

If you're interested in working on it, feel free to claim it.

**Comment by jasonLaster** [comment]
@jfeng21 i believe this is not being worked on. Would you like to try and fix this?

**Comment by fvsch** [comment]
I think @nchevobbe worked on a similar issue for the Console, with a patch reviewed by @yzen.
Perhaps the same solution could apply here, or we could learn something from it?
